### ISSUE TITLE
Allow "track view panning" with middle/any mouse button

### ISSUE THREAD

**Issue by etosan** [audacity/audacity#693]
I don't use audacity often but when I do, I often have to use it in very short stints, or I am risking smashing keyboard or throwing away my mouse :). 

All joking aside, I do believe in what I now call "reorient" principle. Point of this principle is to pan around ever so often using middle mouse button. 

This will be lengthier because of the high level concept involved. First I will describe what I mean by "reorient", and then I will explain how I think it applies to audacity.

I learned to use "reorient" for the first time, when I was working as game developer, and I was taught it by more experienced colleagues, 3D character modellers and animators. When you are working with 3D objects, especially human faces and human bodies, it is essential to work on them in 3D space, because even slightly wrongly positioned vertices break the illusion of the object (face/head). Back then, there were no VR headsets, that would allow for true stereoscopic vision, so monitor was all you've got. Unfortunately monitor is 2D plane projection of your 3D space. 

If you work on 3D model for a longer time (more like 1min) locked to a certain 2D point of view, you might end up moving vertices in such way, that model looks good from that exact 2D POV, but completely wrong from all others, eg. it looks deformed if view gets even slightly rotated. 

Now, some people say that you have 4 views (top, left, right, bottom) into 3D space to avoid that happening, but colleagues taught me immediately first week, that that opinion is bollocks (which it really is). First thing is that these views are in ortographic projection (it means that perspective does not apply to them) and when we look through our eyes, our brains are using perspective projection and second thing is, that humans are not flies, or other insects, and so we don't have four-way vision, and our brains don't work very good with quadviews of 3D models. Whether we want to admit it or not, we look at objects in stereoscopic perspective.

When are using "reorient" principle, you always work in 3D in perspective mode instead of any 2D view, and every 3-30 seconds you **rotate**, **zoom** and **pan** your view back and forth slightly around the object you are editing. These shorts bursts of repositioning and rotation cause you to experience 3D-ness of the object edited even on 2D-viewplane and allows your brain to "read" perspective of the edits and visually check the shape and silhouette of the focused object. This checking is mostly feel based, not science based, and works very well for artistically based edits.

Now how does it apply to audacity? I turns out "reorient" is extremely useful even in 2D views as reorienting allows you to gain better perspective! 

Some years later I worked in completely different field, on database schemas and we used SQL editor with combined SQL REPL and 2D based graphical database diagram view. It turned out, reorient principle worked here perfectly, even on huge 2D database schema canvases. When studying or modifying schemas, I would quickly circle-pan around tables and relations, quickly zooming in and out and this would allow me to notice more of the actual database situation, then to stare dumbly into tiny 2D viewpoint of table I was editing!

This was in 00-10s, I was still using windows for GUI back then. In that timeframe, cca 98-2007, it was customary to have pan bound to middle mouse button in high performance professional windows GUI programs.

Now when I moved to *nix system administration, post 2007, I didn't use GUIs that much. But every time I tried to do anything in *nix GUI software I ended up laughing hysterically, facepalming, or wanting to break keyboard in rage. Almost every major GTK based software has almost completely broken panning on *nix. At first I thought it was only *nix GUI problem but later I found out in the mean time the situation became equally tragic on all platforms. 

The point is that universal panning behavior with "middle mouse button pressed" got fucked up globally. These days even windows programs don't know how to pan with middle mouse and this is getting worse by day. Causes for this are outside of the scope of the issue discussed here, but could be summarized by pandering to lowest common denominator (notebook touchpads, touch tablets, and retarded apple magic mouse).

But, when you are using audacity on your workstation to edit sound, you are not working in some constrained environment. You are using proper equipment like full sized keyboard and proper fully fledged 3-button mouse. 

I think in workstation situation pan mode is not out of question and I believe it would bring new qualities to audacity sound editing.

As such I propose to implement ability to pan view of all tracks while "Middle Mouse" is being held.  Now of course I explored all the possibilities of audacity and I know there is option to use Shift+"Middle Scroll" to pan track view around, but this is not as efficient as "Middle Mouse Hold" pan. This "view pan" mode would be only similar to Shift + "Middle Scroll", as it should be relative to onscreen cursor movement instead of scrollwheel turns. This would combine better Ctrl+"Middle Scroll" for zooming and "Middle Scroll" for scrubbing, for example.

**Comment by SteveDaulton** [comment]
I expect there would be a large scale outcry if we removed "Shift + Mouse wheel" or the horizontal scroll bar, so I assume that you mean that the proposed new way to scroll horizontally would be in addition to "Shift + Mouse wheel" and the horizontal scroll bar?
Personally I'm not convinced that we need a third way to do the same thing, especially, as you say, 3rd button click and drag has fallen out of favour in recent years, and it is largely inaccessible for many Mac and laptop users (but that's just my personal opinion). Historically we have tended to avoid the middle mouse button because many Audacity users don't have a middle mouse button (especially Mac users, though that has slightly improved).
What I think would undoubtedly be a nice new feature, would be for users to modify mouse bindings in a similar way to how key bindings may be modified, but that's a much bigger project than you are suggesting.

**Comment by user58823** [comment]
GIMP: middle click and drag pans canvas
Inkscape: middle click and drag pans canvas
Scribus: middle click and drag pans page
Blender: (outside of the 3D view) middle click and drag pans and scrolls pretty much everything
_Ocenaudio_ (an audio editor with vastly superior UX, which I would use for everything if it had proper multi-track support): middle click and drag pans view

Audacity: middle click and drag does **NOTHING**.

Just one of the many backwards and _unconventional_ things Audacity does that infuriate me to no end. This behavior has been figured out and standardized for centuries now. There is no excuse to not implement this.

So what if some people don't have a middle mouse button? Should the rest of us have a worse experience because of that? It doesn't take away any of the other methods for them to manipulate the view and there is no behavior this would overwrite.

**Comment by SteveDaulton** [comment]
You could add that Firefox previously had middle mouse button drag for scrolling, but that is now off by default.
Also, the current Windows 10 User experience guidelines UWP apps makes no mention of middle mouse button.
It seems to me that middle click was more popular in the days when it was a button, whereas these days it is more commonly a wheel.
As far as possible we like to keep usage the same across all platforms, which means that we sometimes ignore platform specific features.

**Comment by user58823** [comment]
What are you talking about "in the days when it was a button"? If you didn't know, mice with wheels still have a middle button, which you can use by pressing down on the wheel. Even the most basic 'AmazonBasics' one is a 3-button mouse.

**Pretty much all mice used nowadays still have a middle mouse button.**

So why don't you simply consider 'having no middle mouse button' a platform specific "feature" which you can ignore, then?

I use the middle mouse button _countless times everyday_, including panning my view in every other program except Audacity.

This has been standard behavior for way longer than those platforms having removed the mouse button and is purely an enhancement that doesn't take anything away from those without it.

Also, Firefox probably removed it because it could possibly interfere with the _other_ standard action, _opening a link in a new tab_, not because nobody uses middle mouse buttons.

**Comment by etosan** [comment]
@SteveDaulton:

Just to confirm, I am vehemently *against* removing "Shift + Mouse wheel", sorry for not being specific, that is not what I intended.
I am for addition of "sane" panning support "Middle hold" as it would be in audacity parlance, is seems.

I also want add that I agree with @user58823, that when I say "middle mouse button" I of course mean scrollwheel, which is button in itself (somehow this becoming best kept secret of regular mouse operation?).

I am starting to use audacity for some sound editing, something I never did before, and more and more at that, so this little annoyance is frustrating me to no end as well. 

I dived into the menus a bit and I noticed key/mouse bindings, but it seems those are not editable(?), are those configurable at least code wise (I am okay with recompiling my build just to get what I want)?

Sidenote: I have couple of friends with real music production experience (local bands), they all have purchased "real" DAWs (whatever that means) and it was first considerable software investment for them (after instruments of course). Neither are cheap. When I revealed to them, that I plan to use audacity instead, they looked at me funnily and politely explained to me that audacity is probably not the right tool for the job (from sound engineer perspective). I was quite surprised as I prefer to use OSS and read relatively good things about audacity. 

So far I learned how to prepare some test voiceovers in audacity, with their generic instruction adapted to audacity capabilities, but I had roughly same feeling as using gimp vs anything sane (like krita even). I can tolerate quite a lot, and am rather stubborn, but I am becoming more curious: before I invest too much time into this, maybe audacity really is not right tool for the job, despite my stubbornness? I am starting to gather it's rather "out" there?

**Comment by SteveDaulton** [comment]
Key bindings can be customised: https://manual.audacityteam.org/man/keyboard_preferences.html
Currently mouse bindings cannot be customised. I think it would be a useful addition to Audacity if they were, though I don't expect that to be easy to implement.

**Comment by etosan** [comment]
Thanks, if I find time I will look into it a bit.

**Comment by thehappycheese** [comment]
I am here to vote for middle mouse pan, but I am I am glad to see it already made it into 'accepted' feature requests. I cant wait 😄 👍  

I have many closely spaced short samples on a long track. I need fine panning control when zoomed right in, and the scroll wheel feels awful to use for going sideways. And scroll makes my finger hurt for large movements.

**Comment by Lxtharia** [comment]
+1 Bump!
Would love to see this navigation options as it would make things so much more intuitive and handy

**Comment by jez9999** [comment]
Yeah, this should be a thing.  Sigh.  So many things that should be a thing aren't a thing in Audacity, for no good reason.

**Comment by jez9999** [comment]
"Drag to pan" is a bit more than just mapping a mouse button though, it's functionality that needs implementing.

**Comment by gobshyte** [comment]
So I can click on the heading and pan... but not hold mouse to pan... 4 years later :O

**Comment by UrbanCircles** [comment]
any updates on this? seems like such a core UI feature...

**Comment by dozzzzer** [comment]
@UrbanCircles We plan to implement this in Audacity 4.x

**Comment by regs01** [comment]
Control without keyboard (wheel to zoom and middle button to pan) would really greatly improve usability.

**Comment by chinakov** [comment]
This is done, thanks!

**Comment by jez9999** [comment]
Yes!  I'm so happy that Audacity finally has middle click panning, a huge quality of life improvement.

**Comment by regs01** [comment]
Installed 4.0 and it doesn't work. And i don't see it anywhere in Settings either. Was it removed?
